app/controllers/orphans_controller.rb

```diff
@@ -2,54 +2,88 @@ class OrphansController < ApplicationController
   
-  before_action :load_orphans, except: [:index]
+  before_action :load_pseuds, only: [:create]
+  before_action :load_orphans, only: [:create]
 
   def index
     @user = User.orphan_account
     @works = @user.works
   end
 
-  def load_orphans
+  def new
     if params[:work_id]
-      work = Work.find(params[:work_id])
-      @pseuds = (current_user.pseuds & work.pseuds)
-      @orphans = [work]
+      @to_be_orphaned = Work.find(params[:work_id])
+      check_one_owned(@to_be_orphaned, current_user.works)
     elsif params[:work_ids]
-      @orphans = Work.joins(pseuds: :user).where("users.id = ?", current_user.id).where(id: params[:work_ids]).readonly(false)
-      @pseuds = current_user.pseuds & (@orphans.collect(&:pseuds).flatten)
+      @to_be_orphaned = Work.where(id: params[:work_ids]).to_a
+      check_all_owned(@to_be_orphaned, current_user.works)
     elsif params[:series_id]
-      series = Series.find(params[:series_id])
-      @pseuds = (current_user.pseuds & series.pseuds)
-      @orphans = [series]
+      @to_be_orphaned = Series.find(params[:series_id])
+      check_one_owned(@to_be_orphaned, current_user.series)
     elsif params[:pseud_id]
-      @pseuds = [Pseud.find(params[:pseud_id])]
-      @orphans = @pseuds.first.works
+      @to_be_orphaned = Pseud.find(params[:pseud_id])
+      check_one_owned(@to_be_orphaned, current_user.pseuds)
     else
-      @pseuds = current_user.pseuds
-      @orphans = current_user.works
+      @to_be_orphaned = current_user
     end
+  end
 
-    if @pseuds.empty?
-      flash[:error] = ts("You don't have permission to orphan that!")
-      redirect_to root_path and return
-      false
-    end
+  def create
+    use_default = params[:use_default] == "true"
+    Creatorship.orphan(@pseuds, @orphans, use_default)
+    flash[:notice] = ts("Orphaning was successful.")
+    redirect_to(current_user)
   end
 
+  protected
 
-  def new
+  def show_orphan_permission_error
+    flash[:error] = ts("You don't have permission to orphan that!")
+    redirect_to root_path
   end
 
-  def create
-    new_orphans = {}
-    use_default = params[:use_default] == "true"
-    if !@pseuds.blank? && Creatorship.orphan(@pseuds, @orphans, use_default)
-      flash[:notice] = ts('Orphaning was successful.')
-      redirect_to(current_user)
+  # Given an ActiveRecord item and an ActiveRecord relation, check whether the
+  # item is in the relation. If not, show a flash error.
+  def check_one_owned(chosen_item, all_owned_items)
+    show_orphan_permission_error unless all_owned_items.where(id: chosen_item.id).exists?
+  end
+
+  # Given a collection of ActiveRecords and an ActiveRecord relation, check
+  # whether all items in the collection are contained in the relation. If not,
+  # show a flash error.
+  def check_all_owned(chosen_items, all_owned_items)
+    chosen_ids = chosen_items.map(&:id)
+    owned_ids = all_owned_items.where(id: chosen_ids).pluck(:id)
+    unowned_ids = chosen_ids - owned_ids
```

Comment:
I don't think so, because that would return the set of all IDs in `all_owned_items` that aren't in `chosen_items`.  The `unowned_ids` array is supposed to be the reverse: all IDs in `chosen_items` that aren't in `all_owned_items`.

(I could probably do the reverse by making sure that `chosen_items` is a relation, and plucking all IDs from `all_owned_items`. But I was trying to reduce the number of queries.)

Reply:
I missed that, this is good as is.